libyul/ObjectCache.h

```diff
@@ -0,0 +1,84 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libyul/ASTForward.h>
+#include <libyul/Object.h>
+
+#include <liblangutil/EVMVersion.h>
+
+#include <libsolutil/FixedHash.h>
+
+#include <map>
+#include <memory>
+#include <optional>
+
+namespace solidity::yul
+{
+
+// TMP: Move this to a more general place?
+enum class Language
+{
+	Yul,
+	Assembly,
+	StrictAssembly,
+};
+
+Dialect const& languageToDialect(Language _language, langutil::EVMVersion _version);
+
+class ObjectCache
+{
+public:
+	struct Context
+	{
+		Language language;
+		langutil::EVMVersion evmVersion;
+		bool optimizeStackAllocation;
+		std::string yulOptimiserSteps;
+		std::string yulOptimiserCleanupSteps;
+		size_t expectedExecutionsPerDeployment;
+	};
+
+	void optimize(Object& _object, Context const& _context) { optimize(_object, _context, true); }
+
+private:
+	struct CachedObject
+	{
+		std::shared_ptr<Block const> ast;
+		std::shared_ptr<AsmAnalysisInfo> analysisInfo; // TMP: is it really ok to share this without copying?
```

Comment:
Probably not... the analysis info will be used in the code transform and relates pointers to AST elements to pointer abstractions of identifiers, so that should break if you copy one but share the other - that should get you exceptions in the code transform even, shouldn't it? I.e. I'd expect the ``m_info.scopes.at`` accesses there to throw if you try to run the code transform with a copied ast but old analysisInfo.

Asm analysis should be cheap enough to just re-perform in any case.

Reply:
Good to know. But things like these should really be stored `const` in the object and just replaced wholesale with a new thing when they change. Having to copy or recalculate them is really awkward.

I expected that I'll have to make a copy anyway, but for now wanted to see if it'll even work. Apparently it does (I think the failing tests are not failing for a different reason).

Reply:
Yeah, it works due to https://github.com/ethereum/solidity/blob/91b1254c600d712efbd457759168da4c48d82975/libyul/YulStack.cpp#L124 replacing the analysis info anyways afterwards. But yes, making it const or even more strongly enforcing it to be recalculated seems good.